p2p/remote.go

```diff
@@ -0,0 +1,99 @@
+package p2p
+
+import (
+	"context"
+	"gx/ipfs/QmNqRnejxJxjRroz7buhrjfU8i3yNBLa81hFtmf2pXEffN/go-multiaddr-net"
+	ma "gx/ipfs/QmUxSEGbv2nmYNnfXi7839wwQqTN3kwQeUxe8dTjZWZs7J/go-multiaddr"
+	"gx/ipfs/QmXdgNhVEgjLxjUoMs5ViQL7pboAt3Y7V7eGHRiE4qrmTE/go-libp2p-net"
+	"gx/ipfs/QmZNkThpqfVXs9GNbexPrfBbXSLNYeKrE7jwFM2oqHbyqN/go-libp2p-protocol"
+)
+
+var maPrefix = "/" + ma.ProtocolWithCode(ma.P_IPFS).Name + "/"
+
+// remoteListener accepts libp2p streams and proxies them to a manet host
+type remoteListener struct {
+	p2p *P2P
+
+	// Application proto identifier.
+	proto protocol.ID
+
+	// Address to proxy the incoming connections to
+	addr ma.Multiaddr
+}
+
+// ForwardRemote creates new p2p listener
+func (p2p *P2P) ForwardRemote(ctx context.Context, proto protocol.ID, addr ma.Multiaddr) (Listener, error) {
+	listener := &remoteListener{
+		p2p: p2p,
+
+		proto: proto,
+		addr:  addr,
+	}
+
+	if err := p2p.Listeners.Register(listener); err != nil {
+		return nil, err
+	}
+
+	return listener, nil
+}
+
+func (l *remoteListener) start() error {
+	// TODO: handle errors when https://github.com/libp2p/go-libp2p-host/issues/16 will be done
```

Comment:
So, I think this may actually be a problem. As implemented, we can register two "remote" listeners on the same protocol with two different local addresses (they'll have different keys).

However, I doubt it's any worse than it is today.